internal/mesh/internal/controllers/routes/generate.go

```diff
@@ -217,42 +217,53 @@ func compile(
 
 		// Inject catch-all rules to ensure stray requests will explicitly be blackholed.
 		if !top.Default {
-			if !types.IsTCPRouteType(top.RouteType) {
+			if types.IsTCPRouteType(top.RouteType) {
+				if len(top.TCPRules) == 0 {
+					// User requests traffic blackhole.
+					appendDefaultRouteNode(top, types.NullRouteBackend)
+				}
+			} else {
 				// There are no match criteria on a TCPRoute, so never a need
 				// to add a catch-all.
 				appendDefaultRouteNode(top, types.NullRouteBackend)
 			}
 		}
 
-		mc := &pbmesh.ComputedPortRoutes{
-			UsingDefaultConfig: top.Default,
-			Targets:            top.NewTargets,
-		}
 		parentRef := &pbmesh.ParentReference{
 			Ref:  parentServiceRef,
 			Port: port,
 		}
 
+		mc := &pbmesh.ComputedPortRoutes{
+			UsingDefaultConfig: top.Default,
+			ParentRef:          parentRef,
+			Protocol:           allowedPortProtocols[port],
+			Targets:            top.NewTargets,
+		}
+
 		switch {
 		case resource.EqualType(top.RouteType, types.HTTPRouteType):
+			if mc.Protocol == pbcatalog.Protocol_PROTOCOL_TCP {
```

Comment:
GAMMA basically implies that "first xRoute written wins", so in that case whatever protocol the user has defined on the `Service` is not the _effective_ protocol in use when we render the xDS.

We ultimately need the rendered xDS to be self-consistent, so when we have an `HTTPRoute` with an underlying `Protocol` of `"tcp"` we upgrade it when we pass it off to the proxy state template layer.